File: stacks/go/net-http/rest/bff/flow/inference/unclear_ownership.go
package inference

import (
	"idp-go-net-http-rest/bff/flow"
	"sort"
	"strings"
)

func summarizeOwners(hints []flow.NormalizedOwnershipHint) []string {
	owners := make(map[string]struct{})
	for _, hint := range hints {
		for _, team := range hint.OwnerTeamNames {
			if team != "" {
				owners[team] = struct{}{}
			}
		}
		for _, actor := range hint.OwnerActorIDs {
			if actor != "" {
				owners[actor] = struct{}{}
			}
		}
	}
	names := make([]string, 0, len(owners))
	for name := range owners {
		names = append(names, name)
	}
	sort.Strings(names)
	return names
}

func hasConflict(hints []flow.NormalizedOwnershipHint) bool {
	if len(hints) < 2 {
		return false
	}
	seen := make(map[string]struct{})
	for _, hint := range hints {
		key := summarizeOwners([]flow.NormalizedOwnershipHint{hint})
		joined := strings.Join(key, ",")
		if _, ok := seen[joined]; ok && joined != "" {
			continue
		}
		seen[joined] = struct{}{}
	}
	return len(seen) > 1
}

func InferUnclearOwnership(input flow.ProviderAdapterInput) (*flow.FlowSignal, bool) {
	if len(input.OwnershipHints) == 0 {
		return nil, false
	}

	unclear := false
	for _, hint := range input.OwnershipHints {
		if len(hint.OwnerActorIDs) == 0 && len(hint.OwnerTeamNames) == 0 {
			unclear = true
			break
		}
	}

	conflict := hasConflict(input.OwnershipHints)

	if !unclear && !conflict {
		return nil, false
	}

	partial := false
	for _, hint := range input.OwnershipHints {
		if hint.IsPartial {
			partial = true
			break
		}
	}

	owners := summarizeOwners(input.OwnershipHints)

	signal := flow.FlowSignal{
		ID:         "unclear_ownership",
		Title:      "Unclear ownership",
		Severity:   "medium",
		Confidence: degradeConfidence("high", partial),
		Explanation: func() string {
			if conflict {
				return "Conflicting owners detected: " + strings.Join(owners, ", ")
			}
			return "No ownership signals found for this scope."
		}(),
		RecommendedNextAction: "Assign or confirm a single accountable owner for the affected scope.",
		RelatedEntities:      make([]any, 0),
	}
	for _, owner := range owners {
		signal.RelatedEntities = append(signal.RelatedEntities, owner)
	}

	return &signal, true
}

File: stacks/nodejs/react-fastify/rest/bff/src/flow/inference/unclearOwnership.ts
import { degradeConfidence, summarizeOwners } from "./explain";
import type { ProviderAdapterInput, NormalizedOwnershipHint } from "../types";
import type { FlowSignal } from "./types";

function isUnclear(hint: NormalizedOwnershipHint): boolean {
  const owners = (hint.owner_actor_ids ?? []).length + (hint.owner_team_names ?? []).length;
  return owners === 0;
}

function hasConflict(hints: NormalizedOwnershipHint[]): boolean {
  const ownerSets = hints.map((hint) => summarizeOwners([hint]).join(","));
  const unique = new Set(ownerSets.filter((o) => o.length > 0));
  return unique.size > 1;
}

export function inferUnclearOwnership(input: ProviderAdapterInput): FlowSignal | null {
  if (input.ownership_hints.length === 0) {
    return null;
  }

  const unclearHints = input.ownership_hints.filter(isUnclear);
  const conflict = hasConflict(input.ownership_hints);

  if (unclearHints.length === 0 && !conflict) {
    return null;
  }

  const owners = summarizeOwners(input.ownership_hints);
  const scopeDescription = owners.length > 0 ? owners.join(", ") : "no confirmed owners";

  const partialFlag = input.ownership_hints.some((hint) => hint.is_partial);
  const service = input.repository.full_name ?? input.repository.provider_id;
  const [team] = owners;

  return {
    id: "unclear_ownership",
    title: "Unclear ownership",
    severity: "medium",
    confidence: degradeConfidence("high", partialFlag),
    explanation: conflict
      ? `Conflicting owners detected: ${scopeDescription}.`
      : "No ownership signals found for this scope.",
    recommendedNextAction: "Assign or confirm a single accountable owner for the affected scope.",
    relatedEntities: [service, ...owners],
    scope: { service, team, stage: "ownership" },
    observedAt: input.repository.fetched_at,
  };
}
